demos/http/http_demo_s3_download/http_demo_s3_download.c

```diff
@@ -0,0 +1,625 @@
+
+        if( ( *pFileSize == 0 ) || ( *pFileSize == UINT32_MAX ) )
+        {
+            LogError( ( "Error using strtoul to get the file size from %s: fileSize=%d",
+                        pFileSizeStr, ( int32_t ) *pFileSize ) );
+            httpStatus = HTTP_INVALID_PARAMETER;
+        }
+
+        LogInfo( ( "The file is %d bytes long.", ( int32_t ) *pFileSize ) );
+    }
+    else
+    {
+        LogError( ( "Failed to read Content-Range header from HTTP response: Error=%s.",
+                    HTTPClient_strerror( httpStatus ) ) );
+    }
+
+    return httpStatus;
+}
+
+static int downloadS3ObjectFile( const TransportInterface_t * pTransportInterface,
+                                 const char * pHost,
+                                 const char * pMethod,
+                                 const char * pPath )
+{
+    int returnStatus = EXIT_SUCCESS;
+    HTTPStatus_t httpStatus = HTTP_SUCCESS;
+
+    /* The size of the file we are trying to download in S3. */
+    size_t fileSize = 0;
+    /* The number of bytes we want to request with in each range of the file bytes. */
+    size_t numReqBytes = 0;
+    /* curByte indicates which starting byte we want to download next. */
+    size_t curByte = 0;
+
+    /* Initialize the request object. */
+    requestInfo.pHost = pHost;
+    requestInfo.hostLen = strlen( pHost );
+    requestInfo.method = pMethod;
+    requestInfo.methodLen = strlen( pMethod );
+    requestInfo.pPath = pPath;
+    requestInfo.pathLen = strlen( pPath );
+
+    /* Set "Connection" HTTP header to "keep-alive" so that multiple requests
+     * can be sent over the same established TCP connection. This is done in
+     * order to download the file in parts. */
+    requestInfo.reqFlags = HTTP_REQUEST_KEEP_ALIVE_FLAG;
+
+    /* Set the buffer used for storing request headers. */
+    requestHeaders.pBuffer = userBuffer;
+    requestHeaders.bufferLen = USER_BUFFER_LENGTH;
+
+    /* Initialize the response object. The same buffer used for storing
+     * request headers is reused here. */
+    response.pBuffer = fileDownloadBuffer;
+    response.bufferLen = FILE_BUFFER_LENGTH;
+
+    /* Verify the file exists by retrieving the file size. */
+    httpStatus = getS3ObjectFileSize( &fileSize,
+                                      pTransportInterface,
+                                      pHost,
+                                      pMethod,
+                                      pPath );
+
+    if( fileSize < FILE_BUFFER_LENGTH )
+    {
+        numReqBytes = fileSize;
+    }
+    else
+    {
+        numReqBytes = FILE_BUFFER_LENGTH;
+    }
+
+    /* Here we iterate sending byte range requests until the full file has been
+     * downloaded. We keep track of the next byte to download with curByte.
+     * When this reaches the fileSize we stop downloading. */
+    while( curByte < fileSize && httpStatus == HTTP_SUCCESS )
+    {
+        if( httpStatus == HTTP_SUCCESS )
```

Comment:
We can get rid of this first `if( httpStatus == HTTP_SUCCESS )` since it's checked in the while condition